src/concrete/peer.ts

```diff
@@ -0,0 +1,153 @@
+                }
+                const newPeerId: string = Guid.create().toString();
+                message.originalSenderGuid = newPeerId;
+                const body = JSON.parse(message.body);
+
+                const routingTable = this.connectionService.routingTable.clone();
+                routingTable.addPeer(this.connectionService.GUID, THIS_IP);
+
+                const responseBody = JSON.stringify({guid: newPeerId, ip: body.ip,
+                                    routingTable: Array.from(routingTable.peers)});
+
+                // Add the new peer to our registry
+                this.connectionService.routingTable.addPeer(newPeerId, body.ip);
+                
+                await response(new Message(MessageType.JOIN_RESPONSE, newPeerId, responseBody));
+
+                // Let other peers know about the newly joined peer
+                await this.connectionService.sendBroadcast(new Message(MessageType.NEW_PEER, this.connectionService.GUID));
+            }
+        });
+
+        // Handle new peer messages
+        this.connectionService.registerReceiveHandlerForMessageType(MessageType.NEW_PEER, async (message: Message, senderGuid: string) => {
+            if (senderGuid && message.body) {
+                // Add the new peer to our registry
+                const body = JSON.parse(message.body);
+                if (this.connectionService.GUID !== body.guid) {
+                    this.connectionService.routingTable.addPeer(body.guid, body.sender);
+                }
+            }
+        });
+
+        // Handle leave messages
+        this.connectionService.registerReceiveHandlerForMessageType(MessageType.LEAVE, async (message: Message, senderGuid: string) => {
+            if (message && senderGuid) {
+                // Remove the new peer from our registry
+                this.connectionService.routingTable.removePeer(senderGuid);
+            }
+        });
+    }
+    /**
+     * Try to join the network. Send a join request to every given peer.
+     */
+    private checkInitialPeers(peers: string[]): Promise<void> {
+        return new Promise(async (resolve, reject) => {
+            if (!this.connectionService.GUID) {
+                reject();
+                return;
+            }
+            for (const peer of peers) {
+                // Check if peer is online and try to join
+                const message = new Message(MessageType.JOIN, this.connectionService.GUID, JSON.stringify({ip: THIS_IP}));
+                await this.connectionService.sendMessageByIp(message, peer,
+                    async (responseMessage: Message) => { await this.joinResponse(responseMessage); });
+                await this.connectionService.getPromiseForResponse(message);
+            }
+            resolve();
+        });
+    }
+
+    private joinResponse(message: Message): Promise<void> {
+        return new Promise(async (resolve) => {
+            if (message.body && message.originalSenderGuid) {
```

Comment:
Please make sure to use early returns here. Don't forget to apply this into the **entire** project and not just on this line.